### PR TITLE
Integration testing strategy

### LINKED ISSUE #13: Cross: Design Integration Testing Strategy
AS A system engineer
I WANT to understand the current thinking on the overall strategy for integration testing
SO THAT I can write integration tests for the components I am working on

This is the initial draft for the integration testing strategy for all system components. This will only be a draft at this stage because certain aspects are dependent on concurrent work around the overall technical design and architecture. However, we do want to capture what we've agree so far around integration testing.

## Acceptance Criteria
A document which covers:

- [x] Establish meaning of key testing terms in the context of this project - https://github.com/buerokratt/Cross-functional-requirements/pull/61
- [x] Suggested tooling and platform for integration tests- https://github.com/buerokratt/Cross-functional-requirements/pull/61
- [x] Suggested objectives for integration tests- https://github.com/buerokratt/Cross-functional-requirements/pull/61
- [x] How integration tests will be used as a gate for PR's, including integration with CD pipelines- - https://github.com/buerokratt/Cross-functional-requirements/pull/61
- [x] How integration tests work in terms of azure environments- https://github.com/buerokratt/Cross-functional-requirements/pull/61

### PR DESCRIPTION
**Pull request by martinkearn**
This PR adds the integration test strategy document

### CODE REVIEW

File: docs/technical/integration-testing-strategy.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+While this does not give us full assurance that the components integrate in their deployed state, it will enable us to catch integration issues early on.
+
+### Deployed components via infrastructure CD pipeline
+
+In addition to running the integration tests as part of the code repository CI pipeline, the tests will be run again against the deployed version of the code in the Azure Dev environment.
+
+This will use environment data from Terraform to exercise the live components.
+
+Sucessfull completion of the test against the Dev environment will be a gate for the code progressing to the prod environment.
```
Comment (guypritchard) [inline docs/technical/integration-testing-strategy.md:95]:
```suggestion
Successful completion of the test against the Dev environment will be a gate for the code progressing to the prod environment.
```

File: docs/technical/integration-testing-strategy.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+The tests will use an `appsettings.json` and `appsettings.developments.json` files to store configuration information for accessing components (endpoints, api keys etc).
+
+Developers will be able to run tests locally using the `dotnet test` command.
+
+## Pipelines
+
+The goal of running tests as part of pipelines that the outcome of a test can be used a gate for progressing code to the next stage. For example if a test fails, then the automated progression of code from dev to production can be halted.
+
+The challenge with running integration tests as part of a pipeline is that the code must be available in its integrated form in order for the test to work, so to a certain extent the pipline must have progressed before integration tests can run.
```
Comment (guypritchard) [inline docs/technical/integration-testing-strategy.md:65]:
```suggestion
The challenge with running integration tests as part of a pipeline is that the code must be available in its integrated form in order for the test to work, so to a certain extent the pipeline must have progressed before integration tests can run.
```

File: docs/technical/integration-testing-strategy.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+In order to assert that these HTTP requests are being received, we can use some of the existing endpoints that are included in the mock bot to create chats and post messages to a chat (thereby sending it to DMR).
+
+There may be some minor extensions required to mock bot to ensure that when DMR responds back to the bot, the message is added to the Chat (currently it is disposed). This will enable us to simply get the chat and inspect the contents which should contain both the initial message out to DMR and the classified response back from DMR.
+
+## Platform and Tooling
+
+All integration tests will be written using C# and XUnit. This is a consistent choice with the unit tests for the Bürokratt system.
+
+The tests will use an `appsettings.json` and `appsettings.developments.json` files to store configuration information for accessing components (endpoints, api keys etc).
```
Comment (guypritchard) [inline docs/technical/integration-testing-strategy.md:57]: Will this always be the case?  I was wondering if some of the settings (api keys) might come from environment variables when running in a pipeline.  I'm probably misunderstanding here.

File: docs/technical/integration-testing-strategy.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+This section will outline the integration tests that will be created and what assertions they will make. This only outlines the initial "happy path" tests that are required to build assurances that the integrated system is operating correctly.
+
+In passing these tests, all system components (Mock Bot, DMR, Mock Classifier and CentOps) will have been exercised. Additional tests which check edge cases and error cases can be added later.
+
+### Classify a message
+
+This test asserts that when DMR receives an unclassified message which is correctly formed, it is able to classify it and reply to the calling participant with the classification. This is steps 1-4 from [Milestones.md](https://github.com/buerokratt/Project-Documentation-Management/blob/main/Milestones.md) and represents milestone 1.
+
+Input: Test requests that mock bot generates an unclassified message with all appropriate headers and submits it to DMR.
```
Comment (guypritchard) [inline docs/technical/integration-testing-strategy.md:33]: There might be some setup with enrolling the bot(s) needed by tests with CentOps.

**Review by guypritchard — APPROVED** [review]
Nice!  There are some interesting challenges to tackle here.

File: docs/technical/integration-testing-strategy.md (hunk 2 above)
Comment (decodingahmed) [inline docs/technical/integration-testing-strategy.md]: This was a little difficult to read so here my suggestion on how to reword it
```suggestion
The challenge with running integration tests as part of a pipeline is that the code must be deployed to the `dev` environment (our first integrated environment), meaning that it needs to have merged into the `main` branch already.
```

**Review by decodingahmed — CHANGES_REQUESTED** [review]

File: docs/technical/integration-testing-strategy.md (hunk 2 above)
Comment (martinkearn) [reply docs/technical/integration-testing-strategy.md]: I'm not sure if the suggestion is quite right either because we can run the tests with Docker Compose. I'll have a look and come up with some better words.

File: docs/technical/integration-testing-strategy.md (hunk 3 above)
Comment (martinkearn) [reply docs/technical/integration-testing-strategy.md]: Fixed, thanks

File: docs/technical/integration-testing-strategy.md (hunk 4 above)
Comment (martinkearn) [reply docs/technical/integration-testing-strategy.md:35]: Added setup steps.

File: docs/technical/integration-testing-strategy.md (hunk 2 above)
Comment (martinkearn) [reply docs/technical/integration-testing-strategy.md]: I ended up removing the paragraph entirely as it was not really adding anything

**Review by decodingahmed — APPROVED** [review]
